File: src/main/java/jisop/command_line/parse/ParsedArguments.java
package jisop.command_line.parse;

import jisop.MissingArgumentException;
import jisop.domain.Argument;
import jisop.infrastructure.KeyValuePair;

import java.io.OutputStream;
import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.Stream;

/**
 *
 * @author mathieu
 */
public class ParsedArguments {

    /// <summary>
    /// 
    /// </summary>
    /// <param name="parsedArguments"></param>
    public ParsedArguments(Collection<Argument> ArgumentWithOptions,
            Collection<RecognizedArgument> recognizedArguments,
            Collection<UnrecognizedArgument> UnRecognizedArguments) {
        this.argumentWithOptions= ArgumentWithOptions;
        this.recognizedArguments = recognizedArguments;
        this.unRecognizedArguments = UnRecognizedArguments;
    }
    public ParsedArguments(ParsedArguments parsedArguments) {
        recognizedArguments = parsedArguments.recognizedArguments;
        argumentWithOptions = parsedArguments.argumentWithOptions;
        unRecognizedArguments = parsedArguments.unRecognizedArguments;
    }
    public Collection<RecognizedArgument> recognizedArguments;
    public Collection<UnrecognizedArgument> unRecognizedArguments;
    public final Collection<Argument> argumentWithOptions;

    public Collection<Argument> UnMatchedRequiredArguments() {
        return argumentWithOptions.stream()
                .filter(arg->arg.required && !isRecognized(arg))
                .collect(Collectors.toList());
    }

    public Stream<String> invoke() {
        recognizedArguments.stream()
                .filter(arg -> null != arg.argument.action)
                .forEach(arg -> arg.argument.action.accept(arg.value));
        return Stream.of("");
    }

    private boolean isRecognized(Argument arg) {
        return recognizedArguments.stream()
                .anyMatch(a->a.argument.equals(arg));
    }

    public ParsedArguments merge(ParsedArguments parsedMethod) {
        return new MergedParsedArguments(this, parsedMethod);
    }

    public void assertFailOnUnMatched() {
        Collection<Argument> unMatchedRequiredArguments = UnMatchedRequiredArguments();

        if (unMatchedRequiredArguments.size() > 0) {
            throw new MissingArgumentException("Missing arguments",
                    unMatchedRequiredArguments.stream().map(a->a.name).toArray(size->new String[size]));
        }
    }

    public Collection<KeyValuePair<String, String>> RecognizedArgumentsAsPairs() {
        return recognizedArguments.stream()
                .map(r-> r.asKeyValue())
                .collect(Collectors.toList());
    }
}

File: src/main/java/isop/command_line/parse/ParsedArguments.kt
package isop.command_line.parse

import isop.MissingArgumentException
import isop.domain.Argument
import isop.infrastructure.KeyValuePair

import java.util.*
import java.util.stream.Collectors
import java.util.stream.Stream

/**
 *
 * @author mathieu
 */
open class ParsedArguments {
    var recognizedArguments: Collection<RecognizedArgument>
    var unRecognizedArguments: Collection<UnrecognizedArgument>
    val argumentWithOptions: Collection<Argument>


    constructor(ArgumentWithOptions: Collection<Argument>,
                recognizedArguments: Collection<RecognizedArgument>,
                UnRecognizedArguments: Collection<UnrecognizedArgument>) {
        this.argumentWithOptions = ArgumentWithOptions
        this.recognizedArguments = recognizedArguments
        this.unRecognizedArguments = UnRecognizedArguments
    }

    constructor(parsedArguments: ParsedArguments) {
        recognizedArguments = parsedArguments.recognizedArguments
        argumentWithOptions = parsedArguments.argumentWithOptions
        unRecognizedArguments = parsedArguments.unRecognizedArguments
    }

    fun UnMatchedRequiredArguments(): Collection<Argument> {
        return argumentWithOptions
                .filter { arg -> arg.required && !isRecognized(arg) }
                .toList()
    }

    open operator fun invoke(): Stream<String> {
        recognizedArguments.stream()
                .filter { arg -> null != arg.argument.action }
                .forEach { arg -> arg.argument.action!!.accept(arg.value!!) }
        return Stream.of("")
    }

    private fun isRecognized(arg: Argument): Boolean {
        return recognizedArguments.stream()
                .anyMatch { a -> a.argument == arg }
    }

    fun merge(parsedMethod: ParsedArguments): ParsedArguments {
        return MergedParsedArguments(this, parsedMethod)
    }

    fun assertFailOnUnMatched() {
        val unMatchedRequiredArguments = UnMatchedRequiredArguments()

        if (unMatchedRequiredArguments.size > 0) {
            throw MissingArgumentException("Missing arguments",
                    unMatchedRequiredArguments.map { a -> a.name }.toTypedArray())
        }
    }

    fun RecognizedArgumentsAsPairs(): Collection<KeyValuePair<String, String?>> {
        return recognizedArguments
                .map { it.asKeyValue() }
                .toList()
    }
}
